composer/trainer/trainer.py

```diff
@@ -752,7 +776,27 @@ def _train_loop(self) -> None:
             if self.checkpoint_saver and self.checkpoint_saver.should_checkpoint(state=state, event=Event.EPOCH_END):
                 self.checkpoint_saver.save_checkpoint(state=state, seed=self.seed, device=self.device)
 
-    def _train_and_compute_loss(self, use_grad_scaling: bool, batch_num_samples: int):
+    @_dynamic_microbatch_wrapper
+    def _engine_after_dataloader(self):
+        self.engine.run_event(Event.AFTER_DATALOADER)
+
+    @_dynamic_microbatch_wrapper
+    def _compute_training_metrics(self, train_metrics: Metric):
+        """Compute train metrics. Adaptively change microbatch size if enabled to maximize GPU
+        usage.
+
+        Args:
+            train_metrics (Metric): Training metrics
+        """
+        with torch.no_grad():
+            for eval_microbatch in self._train_data_spec.split_batch(self.state.batch, self.state.grad_accum):
+                # TODO: Detect if self.run_event(Event.AFTER_DATALOADER) changes the training
+                # data and if so print a warning that metrics may return unexpected results
+                outputs, targets = self.original_model.validate(eval_microbatch)
+                train_metrics.update(outputs, targets)
```

Comment:
I don't think I'm on board with moving wrapping these methods too. I know that `AFTER_DATALOADER` can OOM if we're using seq_len_warmup, but I don't think fixing that is worth the complexity of added calls to this wrapper. seq_len_warmup can always be fixed by moving the dummy pass into something like `BEFORE_FORWARD`.

Reply:
But thoughts @ajaysaini725 @hanlint ? Essentially the question here is what scope of OOMs should dynamic grad accum attempt to capture - algorithms can potentially cause large memory consumption in almost arbitrary parts of the training loop.

Reply:
Agreed that we don't want to extend this to theese other functions. I'm OK keeping the scope of dynamic GA tight, and then expanding it later.

Reply:
Same - it's not worth adding hacks like this to the trainer code just to enable dynamic grad accum to work in an edge case like this.

My proposal is to do one of:
- Abandon dynamic grad accum for now and just tune grad accum before the start of the training loop (after `Event.INIT` though so it can at least take into account model surgery). In this case, if the user is doing auto grad accum and also has methods that fire on anything other than `Event.INIT`, emit a warning that changing memory usage may break auto grad accum and that they should manually set grad accum in this case,
- Keep dynamic grad accum but emit a warning if any algorithms that fire on something other than `Event.INIT` are being used - in the warning can just say that there are known issues with algorithms that change memory and one should not use auto grad accum in this setting.

For now it's better to have a clean implementation that works in most cases rather than a messy implementation that handles all the known edge cases. Once we have people using this feature, we can come up with better ways of handling the edge cases.